### PR TITLE
fix(validation): correctly handle indirectly hidden questions

### PR DESCRIPTION
**Pull request by winged**
Questions can not only be hidden by evaluating their is_hidden expression
but also when they're part of a subform, where the containing form question
is hidden. We need to check for that case as well, which didn't happen till now.

### CODE REVIEW

File: caluma/form/jexl.py [hunk 1]
```diff
@@ -85,8 +125,20 @@ def is_hidden(self, question):
         all_deps_hidden = bool(deps) and all(
             self.is_hidden(self._question(dep)) for dep in deps
         )
-
-        return all_deps_hidden or self.evaluate(question.is_hidden)
+        if all_deps_hidden:
+            self._cache_is_hidden[question.pk] = True
+            return True
+
+        # Also check if the quesiton is hidden indirectly,
```
Comment (sbor23) [inline caluma/form/jexl.py]: Typo: question

File: caluma/form/tests/test_jexl.py [hunk 2]
```diff
@@ -69,3 +72,85 @@ def test_all_deps_hidden(db, form, document_factory, form_question_factory):
...
+    document_factory,
+    answer_factory,
+    info,
+    fq_is_hidden,
+):
+    # Questions can not only be hidden by evaluating their is_hidden expression
+    # but also if they're part of a subform, where the containing form question
+    # is hidden.
+    #
+    # Showcase form to demonstrate this (f: form, q:question, d:document,a:answer):
```
Comment (sbor23) [inline caluma/form/tests/test_jexl.py]:
```suggestion
    # Showcase form to demonstrate this (f: form, q:question, d:document, a:answer):
```

File: caluma/caluma_form/jexl.py [hunk 3]
```diff
@@ -67,12 +70,49 @@ def _question(self, slug):
...
+        to the given question, and checks if they're hidden.
+
+        If all subforms are hidden where the question shows up,
+        the question shall be hidden as well.
+        """
+
+        doc_form = self.context.get("document").form
+
+        # first, build a list of paths to the question
+        def find_path(fq, acc):
```
Comment (czosel) [inline caluma/caluma_form/jexl.py]: nitpick: I don't know what's meant by `fq` - maybe rather use a longer variable name?

File: caluma/caluma_form/jexl.py [hunk 4]
```diff
@@ -67,12 +70,49 @@ def _question(self, slug):
...
+            parent_fqs = Question.objects.filter(
+                Q(type=Question.TYPE_FORM) | Q(type=Question.TYPE_TABLE),
+                Q(sub_form__questions__pk=fq) | Q(row_form__questions__pk=fq),
+            )
+            res = []
+            if doc_form in fq.forms.all():
+                # found a path - add it to our result list
+                res.append(acc + [fq])
+            for pfq in parent_fqs:
```
Comment (czosel) [inline caluma/caluma_form/jexl.py]: same here

File: caluma/caluma_form/validators.py [hunk 5]
```diff
@@ -195,19 +197,88 @@ def validate(self, document, info, validation_context=None, **kwargs):
...
+
+        q_jexl = jexl.QuestionJexl(validation_context)
+        for question in validation_context["questions"].values():
+            try:
+                is_hidden = q_jexl.is_hidden(question)
+
+                if not is_hidden:
+                    visible_questions.append(question.slug)
+                    if question.type == Question.TYPE_FORM:
+                        # form questions's answers are still in the top level document
```
[the remark this answers is not in the document]
Comment (czosel) [reply caluma/caluma_form/validators.py]: :+1:

File: caluma/caluma_form/jexl.py (hunk 3 above)
Comment (winged) [reply caluma/caluma_form/jexl.py]: `question` would be the obvious choice, but that's already present in the outer context. I'll think of something

File: caluma/caluma_form/jexl.py [hunk 6]
```diff
@@ -67,12 +77,59 @@ def _question(self, slug):
             )
         return question
 
+    def _paths_to_question(self, question, doc_form):
+        """Return all paths from the doc form to the given question."""
+        cache = self._cache["containers_hidden_paths"]
+        if question.slug in cache:
+            return cache[question.slug]
```
Comment (czosel) [inline caluma/caluma_form/jexl.py]: The return value of the function seems to depend on `doc_form` but the caching is only based on `question`. Doesn't this cause issues?

Comment (winged) [reply caluma/caluma_form/jexl.py]: There can only be one answer per question in the whole document structure. The only place where that's not true is in table rows. But the "containing question" thing falls down there anyway, as you cannot refer to a specific row anyhow, so the expression cannot be correctly evaluated in that case.

Comment (czosel) [reply caluma/caluma_form/jexl.py]: Hmm... isn't `doc_form` always the main form? if so, then adding it to the caching keys should not change the cache hit ratio, right?
Or are you calling this with for table row documents as well?

Comment (winged) [reply caluma/caluma_form/jexl.py]: Changed the code to include `doc_form` in the cache key

File: caluma/caluma_form/validators.py [hunk 7]
```diff
@@ -195,19 +198,89 @@ def validate(self, document, info, validation_context=None, **kwargs):
...
         if not validation_context:
             validation_context = self._validation_context(document)
+        visible_questions = []
+
+        q_jexl = jexl.QuestionJexl(validation_context)
+        for question in validation_context["questions"].values():
+            try:
+                is_hidden = q_jexl.is_hidden(question)
+
+                if not is_hidden:
```
Comment (sbor23) [inline caluma/caluma_form/validators.py]: You could remove nesting if you do this instead:
```python
if is_hidden:
    continue
# do the validation now ...
```

File: caluma/caluma_form/validators.py [hunk 8]
```diff
@@ -285,67 +358,55 @@ def _validate_required(self, validation_context):  # noqa: C901
...
+                if question.type == Question.TYPE_FORM:
+                    # form questions's answers are still in the top level document
+                    sub_context = {
+                        **validation_context,
+                        "questions": {
+                            q.slug: q for q in question.sub_form.questions.all()
+                        },
+                    }
+                    self._validate_required(sub_context)
+                elif question.type == Question.TYPE_TABLE and is_required:
```
Comment (sbor23) [inline caluma/caluma_form/validators.py]: Just a question: Why do we validate TableQuestions only if `is_required`, but FormQuestions are validates anyway?

Comment (czosel) [reply caluma/caluma_form/validators.py]: I think FormQuestions don't have a real required state on their own - it's given by the requiredness of the questions the form contains.

**Comment by czosel** [comment]
Coverage is still lacking: https://travis-ci.com/projectcaluma/caluma/jobs/277914328

File: caluma/caluma_form/validators.py (hunk 8 above)
Comment (sbor23) [reply caluma/caluma_form/validators.py]: Ah makes sense, thanks

**Comment by winged** [comment]
> Coverage is still lacking: https://travis-ci.com/projectcaluma/caluma/jobs/277914328

Nice, missing coverage pointed out a refactoring/cleanup bug! (fixed)

**Comment by czosel** [comment]
Awesome! Now let's make sure @open-dynaMIX 's feedback is addressed and then we can get this merged :rocket:

**Comment by winged** [comment]
As discussed, let's merge this first, then do the 5.0 release

**Pull request merged**